ocaml/typing/solver.ml

```diff
@@ -33,17 +33,37 @@ type 'a error =
     right : 'a
   }
 
-(** Map the function to the list, and returns the first [Error] found;
-    Returns [Ok ()] if no error. *)
-let rec find_error (f : 'x -> ('a, 'b) Result.t) : 'x list -> ('a, 'b) Result.t
-    = function
-  | [] -> Ok ()
-  | x :: rest -> (
-    match f x with Ok () -> find_error f rest | Error _ as e -> e)
-
 module Solver_mono (C : Lattices_mono) = struct
+  type any_morph = Any_morph : ('a, 'b, 'd) C.morph -> any_morph
+
+  module VarMap = Map.Make (struct
+    type t = int * any_morph
+
+    let compare (i1, m1) (i2, m2) =
+      match Int.compare i1 i2 with 0 -> compare m1 m2 | i -> i
+  end)
+
+  (** Map the function to the list, and returns the first [Error] found;
+      Returns [Ok ()] if no error. *)
+  let find_error (f : 'x -> ('a, 'b) Result.t) (t : 'x VarMap.t) :
+      ('a, 'b) Result.t =
+    let r = ref (Ok ()) in
+    let _ =
+      VarMap.for_all
+        (fun _ x ->
+          match f x with
+          | Ok () -> true
+          | Error _ as e ->
+            r := e;
+            false)
+        t
+    in
+    !r
+
+  let var_map_to_list t = VarMap.fold (fun _ a xs -> a :: xs) t []
+
   type 'a var =
-    { mutable vlower : 'a lmorphvar list;
+    { mutable vlower : 'a lmorphvar VarMap.t;
```

Comment:
I'm slightly sad that this is a map instead of set. I can see the circular dependency that caused this. I wonder if we should use some `Obj.magic` to workaround the circle and improve performance.

Reply:
I would hope there's a way of doing this without `Obj.magic`. Maybe some recursive modules? But it's probably not worth going through extreme measures. I wouldn't expect a set to be much (or at all, really) faster than a map.

Reply:
Yeah I was also wondering about memory savings. But yeah, magic is maybe not worthwhile.